rfcs/000-caller-side-managed-buffers.md

```diff
@@ -0,0 +1,833 @@
+
+```rust
+pub struct IncomingTls<B> {
+    buf: B,
+
+    filled: usize,
+
+    // keeps track of how many bytes have already been discarded from the first app-data record
+    // - `None` indicates that the record in the front has not yet been decrypted in place
+    // - `Some` indicates that the record in the front has been decrypted in place; the inner
+    //   value indicates how many bytes have already been discarded
+    partially_discarded: Option<usize>,
+}
+```
+
+As described in the code comment, when `discard_app_data` results in partially discarding a record,
+the `partially_discarded` field is set to the `Some` variant.
+The `decrypt_incoming` iterator will read the `partially_discarded` field and
+only yield the part of the record that has not been discard.
+
+The `decrypt_incoming` iterator also checks the variant of `partially_discarded` as an indicator of
+whether the first record in `incoming_tls` has already been decrypted in place or not.
+When `partially_discarded` is `None`, the first record is expected to be encrypted;
+when it's `Some`, the expectation is that the record has already been decrypted.
+
+Because the `decrypt_incoming` iterator decrypts a record each time `next` is called,
+with the suggested approach, this can result in a runtime error in the following scenario:
+
+```rust
+let mut records = conn.decrypt_incoming(&mut incoming_tls);
+drop(records.next().unwrap()?);
+
+// forgot to call this!
+// incoming_tls.discard_app_data();
+
+let mut records = conn.decrypt_incoming(&mut incoming_tls);
+assert!(records.next().unwrap().is_err());
+```
+
+Basically, every record decrypted using `next()` must be discarded, or partially discarded, prior to using a new instance of the `decrypt_incoming` iterator or that iterator will eventually error on a `next` call when it sees a decrypted record when a encrypted one was expected.
+
+There are ways to address this last issue: `IncomingTls` could also store the numbers of records that have already been decrypted in place.
+`partially_discarded` would still refer to the first record.
+It's unclear if decrypting records that won't be immediately used should be supported or encouraged.
+At first glance, it seems wasteful but "decrypting ahead of time" could have some use case.
+
+## Misusing the API
+
+This section describes some scenarios where the API is misused and what the outcomes are.
+
+- `Status.state` is `NeedsMoreTlsData`, no data or an incomplete TLS record is appended to `incoming_tls` and `process_tls_records` is called
+  - the newly returned `Status.state` will also be in the `NeedsMoreTlsData` state. the handshake won't make progress until sufficient data is appended to `incoming_tls`
+
+Also see the 'unresolved questions' section.
+
+# Supporting async certificate verification
+
+In the API described thus far the certificate verification still happens implicitly in `process_tls_records`.
```

Comment:
The idea of exposing certificate verification input via `Status`/`State` makes sense to me.